src/app/services/dataset/dataset_processor_service.py

```diff
@@ -0,0 +1,158 @@
+import uuid
+
+from config.config_factory import ConfigFactory
+from logging_config import logging
+from models.dataset_models import (
+    DatasetMetadataWithoutId,
+    NewDatasetMetadata,
+    NewDatasetWithMetadata,
+    UnitDataset,
+)
+from repositories.dataset_repository import DatasetRepository
+from services.dataset.dataset_writer_service import DatasetWriterService
+from services.datetime_service import DatetimeService
+
+config = ConfigFactory.get_config()
+
+
+logger = logging.getLogger(__name__)
+
+
+class DatasetProcessorService:
+    def __init__(self) -> None:
+        self.dataset_repository = DatasetRepository()
+
+        self.dataset_writer_service = DatasetWriterService(self.dataset_repository)
+        pass
+
+    def process_new_dataset(
+        self, filename: str, new_dataset: NewDatasetWithMetadata
+    ) -> None:
+        """
+        Processes the incoming dataset.
+
+        Parameters:
+        filename (str): the filename of the json containing the dataset data
+        new_dataset (NewDatasetWithMetadata): new dataset to be processed
+        """
+        logger.info("Processing new dataset...")
+        logger.debug(f"Dataset being processed: {new_dataset}")
+
+        new_dataset_unit_data_collection = new_dataset.pop("data")
+
+        logger.info("Transforming new dataset metadata...")
+        transformed_dataset = self._transform_new_dataset_metadata(
+            new_dataset, filename, new_dataset_unit_data_collection
+        )
+        logger.info("Dataset transformed successfully.")
+        logger.debug(f"Transformed dataset: {transformed_dataset}")
+
+        logger.info("Writing transformed dataset to repository...")
+        dataset_id = str(uuid.uuid4())
+        self.dataset_writer_service.write_transformed_dataset_to_repository(
+            dataset_id,
+            transformed_dataset,
+        )
+        logger.info("Transformed dataset written to repository successfully.")
+
+        logger.info("Transforming unit data collection...")
+        transformed_unit_data_collection = self._transform_dataset_unit_data_collection(
+            dataset_id, transformed_dataset, new_dataset_unit_data_collection
+        )
+        logger.info("Unit data collection transformed successfully.")
+        logger.debug(
+            f"Transformed unit data collection: {transformed_unit_data_collection}"
+        )
+
+        logger.info("Writing transformed unit data to repository...")
+        self.dataset_writer_service.write_transformed_unit_data_to_repository(
+            dataset_id, transformed_unit_data_collection
+        )
+        logger.info("Transformed unit data written to repository successfully.")
+
+    def _transform_new_dataset_metadata(
```

Comment:
Could be worth renaming to something along the line of adding metadata to dataset or something

Reply:
Yeah good point, I've renamed the unit data one below as well as it's also just adding metadata. Think I wanted to name this function to be generic in case business requirements changed but I think you're right it's better to be explicit